### ISSUE TITLE
Enable Leibniz equality types

### ISSUE THREAD

**Issue by winitzki** [dhall-lang/dhall-lang#1379]
For people who already know the relationship between `assert : a ≡ b` and the Leibniz equality, let me just get to the point and say that I am proposing to enable the following new typing judgment in Dhall typechecker:

```dhall
(x : a ≡ b) : ∀(f : T → Type) → f a → f b
```
where `T` is the type of `a` and `b`. This typing judgment will holds only if `a` and `b` have the type `T` and only if `x : a ≡ b` is already well-typed.

This typing judgment converts Dhall's built-in type `a ≡ b` into an explicit Leibniz equality type.

In the mathematical notation, the new typechecking rule will be something like this:

```
Γ ⊢ T : Type
Γ ⊢ a : T
Γ ⊢ b : T
Γ ⊢ x : a ≡ b
────────────────────────────────────
Γ ⊢ x : ∀(f : T → Type) → f a → f b
```

And the corresponding beta-reduction rule for applying such `x` to arguments:

```
x ⇥ assert : a ≡ b
p₀ ⇥ p₁
──────────────────────
x f p₀ ⇥ p₁
```

(Perhaps these proposed rules need corrections, I'm not an expert in writing that.)

In my view, this is a missing link that prevents Dhall from using `assert` and equality types for dynamic validation.

This new typing judgment will not break or modify any existing Dhall code.
But it will open doors for using equality types with more flexibility; and I believe this is the only functionality that Dhall still needs in that space.

Some more details and motivation:

At the moment, Dhall implements equality types `a ≡ b` and `assert` expressions. This gives us unit testing and constraints that are checked at compile time. But actually this is a form of Leibniz equality in disguise. See the discussion here: https://discourse.dhall-lang.org/t/initial-objects-or-the-empty-map/662

One of the main points in that discussion is that the equality type `a ≡ b` _can be encoded_ in Dhall without need for special syntax, but then the usage would become

> less ergonomic. This is why ≡ is defined to be a language built-in in Dhall rather than implemented as Leibniz equality in the Prelude.

I argue that the implementation of `a ≡ b` in Dhall has a missing feature: one cannot derive Leibniz equality from `a ≡ b`. (But one can derive `a ≡ b` from Leibniz equality.)

That's why I am proposing to add a typing judgment and a beta-reduction rule that enables converting `a ≡ b` values into Leibniz equality values.

Here is an implementation of Leibniz equality that has the same expressive power as Dhall's `a ≡ b`:

```dhall
let LeibnizEqual = λ(T : Type) → λ(a : T) → λ(b : T) → ∀(f : T → Type) → f a → f b
```
To encode a value similar to Dhall's `let x = assert : 123 ≡ 123`, we write:

```dhall
let x : LeibnizEqual Natural 123 123 = λ(f : Natural → Type) → λ(p : f 123) → p
```

For convenience, we can implement a function (usually called "refl"):

```dhall
let refl : ∀(T : Type) → ∀(a : T) → LeibnizEqual T a a = λ(T : Type) → λ(a : T) → λ(f : T → Type) → λ(p : f a) → p
let x : LeibnizEqual Natural 123 123 = refl Natural 123
```

If we want to require that `a` equals `b` at compile time (i.e., at Dhall's type-checking time), we can write code like this:

```dhall
let a : Natural = ./import_some_number_from_file.dhall
let b : Natural = ./import_another_number.dhall
let x : LeibnizEqual Natural a b = refl Natural a  -- This will not type-check unless a equals b.
```

Apart from reflexivity (`refl`), standard properties include function extensionality (if `g = h` and `a = b` then `g a = h b`):

 `LeibnizEqual T a b → LeibnizEqual (T → U) g h → LeibnizEqual U (g a) (h b)`

symmetry (if `a = b` then `b = a`):

`LeibnizEqual T a b → LeibnizEqual T b a`

and transitivity (if `a = b` and `b = c` then `a = c`):

`LeibnizEqual T a b → LeibnizEqual T b c  → LeibnizEqual T a c`

These standard combinators provide "equational reasoning" at the level of evidence values.

If Dhall did not have `assert : a ≡ b`, we could implement equality types via `LeibnizEqual`.
We would then create values of equality types via `refl`.

For example, here is a function that divides `x` by `y` and requires evidence that `y` is nonzero:

```dhall
let Nonzero = λ(x : Natural) → LeibnizEqual Bool (Natural/isZero x) False
let safe_divide_by = λ(y : Natural) → λ(evidence : Nonzero y) → λ(x : Natural) → ...
```
The function `safe_divide_by` can assume that `y` is nonzero (otherwise the type `Nonzero y` is void and no value of type `Nonzero y` could be given).

For a literal nonzero number, we would create a value of type `LeibnizEqual Bool False False` as `refl Bool False` and pass it to the function `safe_divide_by`.

```dhall
let nonzero_123 : Nonzero 123 = refl Bool False
let result = safe_divide_by 123 nonzero_123 246  -- result is 2
```

So, let us list the features of equality types that we need:

1. Encode equality types, create explicit evidence values of those types.
2. Provide reasoning about equations (reflexivity, symmetry, transitivity, function extensionality).
3. Ensures that an equation holds for literal values known at compile time.

Dhall's `assert` can do (1) and (3) but not (2), except for reflexivity. The encoding `LeibnizEqual` can do all three but is more verbose.

Given that Dhall already has `assert` and `a ≡ b`, how would we implement item (2)? I see two possibilities:

- Add three more built-ins to implement symmetry, transitivity, and function extensionality directly for values of types `a ≡ b`.
- Add type-checking rules and beta-reduction rules so that evaluation takes into account the presence of values of types `a ≡ b` and automatically substitutes `a` instead of `b` everywhere.
- Add just one built-in type conversion from `a ≡ b` to `LeibnizEqual`. This is my preferred proposal.

Note that `LeibnizEqual` can be converted into `a ≡ b` directly:

```dhall
let toAssertType = λ(T : Type) → λ(a : T) → λ(b : T) → λ(x : LeibnizEqual T a b) → 
  let Equals_a = λ(x : T) → a ≡ x
  let refl_a = assert : a ≡ a
  let result : a ≡ b = x Equals_a refl_a
    in result
```

But `a ≡ b` cannot be converted into `LeibnizEqual`. So, that's why I am proposing to enable that.

**Comment by Gabriella439** [comment]
Yeah, I think this is a great idea and I'd approve a change like this.

I think my only suggestion is to make the conversion between `a ≡ b` and `∀(f : T → Type) → f a → f b` explicit (e.g. by providing a `asLeibnizEquality` built-in or similar) instead of implicit.

**Comment by winitzki** [comment]
After some thinking, I believe there is a problem converting the type `a ≡ b` to the type `∀(f : T → Type) → f a → f b`.  We can implement `asLeibnizEquality` with minimal changes to the Dhall type-checker only if we restrict its usage to actual asserted values of type `a ≡ b`. In other words: We can easily implement `asLeibnizEquality (assert : a ≡ b)` as I already described earlier in this issue. But it is quite a lot of work to implement `asLeibnizEquality eq` where `eq` is a free variable of type `a ≡ b`.

The problem is that the typechecker needs to take into account that having an argument of type `a ≡ b` means `a` and `b` have the same normal forms. Otherwise the typechecker won't be able to see that the types `f a` and `f b` are the same. But currently the typechecker has no way of knowing about that. The typechecker would need to have a "value context" where it gathers equations such as `a ≡ b` that are known to hold in the current scope. (Currently, the typechecker only has a "type context" where it gathers known type judgments for free variables.)

Here is an example of code that will not typecheck:

```dhall
let f : Bool → Type = λ(b : Bool) → if b then Natural else Text -- whatever
in
λ(a : Bool) → λ(b : Bool) → λ(evidence : a ≡ b) → λ(fa : f a) →
  (asLeibnizEquality evidence f fa) : f b
```

The expression `asLeibnizEquality evidence f fa` must produce a term of type `f b`. What term would that be? That term could only be `fa` itself, because there is no way of producing any other terms of type `f b`. We know that `fa : f a`, and we know that `a ≡ b`, so `fa : f b` should typecheck. But the current typechecker has no way of deriving the type judgment `fa : f b` because `a` and `b` are bound variables and have different normal forms. It would be a lot of work to enable that kind of typechecking.

On the other hand, it is possible to have code like this:

```dhall
let a : Bool = ...
let b : Bool = ...
let f : Bool → Type = λ(b : Bool) → if b then Natural else Text
let assertion = assert : a ≡ b
in
λ(fa : f a) → (asLeibnizEquality assertion f fa) : f b
```
The term `asLeibnizEquality assertion f fa` is just `fa`, and the typechecker will accept the typing judgment `fa : f b` because `a` and `b` are free variables that actually have the same normal forms (or else the typechecker will fail earlier in that code).

So, I'd suggest for now to implement just this functionality: `asLeibnizEquality` for literal assertions.

Please let me know if my thinking is correct.

**Comment by Gabriella439** [comment]
What's the intended use case for this?  The reason I'm asking is that I want to make sure that if we restrict `asLeibnizEquality` to literal `assert`s that it's addressing the original problem
